### PR TITLE
Fix msvc build for client tools

### PR DESCRIPTION
**Pull request by pf-qiu**
Postgres has its own build system for Windows using MSVC. We can
reuse it to build client tools on windows.

- Add 'fullbuild' option to the default config file. By default,
it is set to 0. Only client tools will be built.
- Only generate project file for client tools
- Add detection for MSVC SDK version. If this field is missing,
8.1 is assumed. The latest compiler will complain about this.
- Add glob.c to libpgport, used by gpfdist
- Only install client tools

### CODE REVIEW

**Comment by pf-qiu** [comment]
@hlinnaka @jchampio As we discussed in https://github.com/greenplum-db/gpdb/pull/6657, we will split client tools into two parts.
This PR will fix the build script for psql, pg_dump, libpq using the upstream build system.
We will build gpfdist and pygresql separately. Can you please have a look, thanks.

File: src/port/glob.c [hunk 1]
```diff
@@ -85,6 +85,7 @@
 #include <pwd.h>
 #include <unistd.h>
 #endif
+#include <windows.h>
```
Comment (hlinnaka) [inline src/port/glob.c]: This should be in an #ifdef WIN32 block.

I wonder, why is this needed at all, though? There shouldn't be anything windows-specific in this file.

File: src/tools/msvc/Install.pm [hunk 2]
```diff
@@ -107,13 +107,22 @@ sub Install
...
 		"$conf\\",
-		"postgres\\postgres.lib",
 		"libpq\\libpq.lib",
-		"libecpg\\libecpg.lib",
 		"libpgcommon\\libpgcommon.lib",
-		"libpgport\\libpgport.lib",
-		"libpgtypes\\libpgtypes.lib",
-		"libecpg_compat\\libecpg_compat.lib");
+		"libpgport\\libpgport.lib");
+
```
Comment (hlinnaka) [inline src/tools/msvc/Install.pm]: Can you submit this addition of "install client" to the upstream, please? It makes a lot of sense to have "install client", when we already have "build client"

File: src/tools/msvc/MSBuildProject.pm [hunk 3]
```diff
@@ -34,10 +34,16 @@ sub WriteHeader
...
+	my $sdkversion=$ENV{'WindowsSDKVersion'};
+	if ($sdkversion =~ /.*\\$/)
+	{
+		chop $sdkversion;
+	}
 	print $f <<EOF;
   </ItemGroup>
   <PropertyGroup Label="Globals">
     <ProjectGuid>$self->{guid}</ProjectGuid>
+    <WindowsTargetPlatformVersion>$sdkversion</WindowsTargetPlatformVersion>
```
Comment (hlinnaka) [inline src/tools/msvc/MSBuildProject.pm]: Doesn't the system pick up the right target automatically? Could you set some environment variable that's understood directly by MSVC, without this? The upstream documentation talks about using `setenv` for targeting particular architectures, is there a `setenv` option for this, too?

File: src/tools/msvc/MSBuildProject.pm [hunk 4]
```diff
@@ -307,6 +313,7 @@ sub WriteItemDefinitionGroup
       <SuppressStartupBanner>true</SuppressStartupBanner>
       <DebugInformationFormat>ProgramDatabase</DebugInformationFormat>
       <CompileAs>Default</CompileAs>
+      <MultiProcessorCompilation>true</MultiProcessorCompilation>
```
Comment (hlinnaka) [inline src/tools/msvc/MSBuildProject.pm]: In PostgreSQL version 10 and above, you can setting `$ENV{MSBFLAGS}="/m";` in the `config.pl` file (see https://www.postgresql.org/docs/10/install-windows-full.html). I don't think we should hardcode this here.

File: src/tools/msvc/config_default.pl [hunk 5]
```diff
@@ -21,7 +21,10 @@
 	xml     => undef,    # --with-libxml=<path>
 	xslt    => undef,    # --with-libxslt=<path>
 	iconv   => undef,    # (not in configure, path to iconv)
-	zlib    => undef     # --with-zlib=<path>
+	zlib    => undef,    # --with-zlib=<path>
+
+	fullbuild  => 0  # 0: only build client tools
```
Comment (hlinnaka) [inline src/tools/msvc/config_default.pl]: You can do `build client` to build just the client tools, no need for this new option.

File: src/port/glob.c (hunk 1 above)
Comment (pf-qiu) [reply src/port/glob.c]: Ah, sure. It's under a #ifndef block, we can conveniently use that. I should document this.
The problem lies in win32 port of dirent.h. MAX_PATH macro needs a definition.

Comment (hlinnaka) [reply src/port/glob.c]: Probably best to use MAXPGPATH instead..

File: src/tools/msvc/Install.pm (hunk 2 above)
Comment (pf-qiu) [reply src/tools/msvc/Install.pm]: Sure. I think we need a definition of "client" component first.
From our perspective the list contains psql, libpq, pg_dump and scripts.
Not sure if this is a full list.

File: src/tools/msvc/MSBuildProject.pm (hunk 3 above)
Comment (pf-qiu) [reply src/tools/msvc/MSBuildProject.pm]: Looks like this problem also exists in upstream postgres.
I'm using the MSVC shell so any environment should be there already.
This might be an issue of the MSVC tool chain, to assume 8.1 as the default value.
And the 8.1 SDK does not come with the Windows OS by default.
So if you install the latest toolchain(VS2017) from a fresh machine, you only have 10.0 SDK.
Then the compiler will complain about missing SDK component using the generated project files.

File: src/tools/msvc/config_default.pl (hunk 5 above)
Comment (pf-qiu) [reply src/tools/msvc/config_default.pl]: The upstream build script doesn't have a client target. The install script has.
So I think the logic is to compile all and install some.
But we can't compile all GPDB so I think this is necessary.

Comment (hlinnaka) [reply src/tools/msvc/config_default.pl]: Oh, right, I mixed up the build and install stages.

For symmetry with `install client`, it would be good to also have `build client`. That's what I meant above as well, when I said that it would be good to submit a patch for that to upstream.

**Comment by pf-qiu** [comment]
> On 10/01/2019 11:19, Peifeng Qiu wrote: pf-qiu commented on this pull request. > print $f <<EOF; </ItemGroup> <PropertyGroup Label="Globals"> <ProjectGuid>$self->{guid}</ProjectGuid> + <WindowsTargetPlatformVersion>$sdkversion</WindowsTargetPlatformVersion> Looks like this problem also exists in upstream postgres. I'm using the MSVC shell so any environment should be there already. This might be an issue of the MSVC tool chain, to assume 8.1 as the default value. And the 8.1 SDK does not come with the Windows OS by default. So if you install the latest toolchain(VS2017) from a fresh machine, you only have 10.0 SDK. Then the compiler will complain about missing SDK component using the generated project files.
> Which SDK should we target? If we build with 10.0 SDK, do the binaries work on older Windows versions? In any case, seems easiest to just install the 8.1 SDK on the build machine. You just need to tick a box in the Visual Studio Installer to get it, right? - Heikki

We would like to use the latest SDK. No matter which version we use to compile, end user needs to install the correspond "Visual C++ Redistributable Packages" to run.
If you compile with 10, you will need to install redist v10 on win 8.1.
If you compile with 8.1, you will need to install redist v8.1 on win 10.

I think this change will be a plus to upstream postgres as it would allow pg to be built with any SDK from 8.1 on.

File: src/tools/msvc/config_default.pl (hunk 5 above)
Comment (pf-qiu) [reply src/tools/msvc/config_default.pl]: I see. I will refactor this to make build consistent with install script.

**Comment by berlin-ab** [comment]
@pf-qiu  Looks like this is ready for another round of feedback. Is that correct? cc @hlinnaka @jchampio

**Comment by pf-qiu** [comment]
We still need some work on this. We will submit three other PRs to build gpfdist, pygresql and make the msi package. Along with this one shall be all scripts to build client tools on windows.

**Comment by berlin-ab** [comment]
I labeled the PR as "wip: feedback encouraged - not done yet". Be sure to change the label when you're ready to merge.

**Comment by berlin-ab** [comment]
What's the latest here?

**Comment by pf-qiu** [comment]
We will do the follow up works before GA. Currently we are not focusing on this item. Or we can close it right now and open another when we are ready.

**Comment by pf-qiu** [comment]
Related PR: https://github.com/greenplum-db/gpdb/pull/7514

**Comment by pf-qiu** [comment]
Related PR: https://github.com/greenplum-db/gpdb/pull/7515

**Comment by pf-qiu** [comment]
Related PR: https://github.com/greenplum-db/gpdb/pull/7516
